scripts/testing/programs/match_template/test_template_reproducibility.py

```diff
@@ -0,0 +1,265 @@
+
+        # Verify all threshold values are the same
+        if threshold_values:
+            if not all(abs(v - threshold_values[0]) < 1e-10 for v in threshold_values):
+                print("Warning: Threshold values differ between replicates:")
+                for i, val in enumerate(threshold_values):
+                    print(f"  Replicate {i+1}: {val:.6e}")
+
+            # Use the first threshold value for calculations
+            threshold_value = threshold_values[0]
+        else:
+            print("Warning: No threshold values were extracted.")
+            threshold_value = None
+
+        # Compute similarity metrics between replicates
+        print("\nReproducibility Analysis:")
+        print("========================")
+
+        # Pairwise comparisons of available replicates
+        pairs = [(i, j) for i in range(len(mip_data)) for j in range(i+1, len(mip_data))]
+
+        all_correlations = []
+        all_mean_abs_diffs = []
+        all_psnrs = []
+
+        for i, j in pairs:
+            try:
+                # Calculate correlation coefficient
+                correlation = np.corrcoef(mip_data[i].flatten(), mip_data[j].flatten())[0,1]
+                all_correlations.append(correlation)
+
+                # Calculate mean absolute difference
+                mean_abs_diff = np.mean(np.abs(mip_data[i] - mip_data[j]))
+                all_mean_abs_diffs.append(mean_abs_diff)
+
+                # Calculate relative error if threshold value is available
+                if threshold_value and threshold_value > 0:
+                    relative_error = mean_abs_diff / threshold_value
+                    relative_error_ppm = relative_error * 1e6  # Parts per million
+                else:
+                    relative_error = None
+                    relative_error_ppm = None
+
+                # Calculate peak signal to noise ratio (PSNR)
+                mse = np.mean((mip_data[i] - mip_data[j]) ** 2)
+                max_pixel = max(np.max(mip_data[i]), np.max(mip_data[j]))
+                psnr = 20 * np.log10(max_pixel) - 10 * np.log10(mse) if mse > 0 else float('inf')
+                all_psnrs.append(psnr)
+
+                # Print results for this pair
+                print(f"Comparing replicate {i+1} vs {j+1}:")
+                print(f"  Correlation coefficient: {correlation:.6f}")
+                print(f"  Mean absolute difference: {mean_abs_diff:.6f}")
+
+                # Print relative error if available
+                if relative_error is not None:
+                    print(f"  Relative error (scientific): {relative_error:.6e}")
+                    print(f"  Relative error (ppm): {relative_error_ppm:.2f} ppm")
+
+                print(f"  Peak signal-to-noise ratio (PSNR): {psnr:.2f} dB")
+
+            except Exception as e:
+                print(f"Error comparing replicates {i+1} and {j+1}: {str(e)}")
+
+        # Overall similarity across all replicates
+        if all_correlations:
```

Comment:
Before printing this section, let's print out how many replicates were counted. This could be calculate from the variable defining the number of replicates, or tracked using a counter that is incremented when running the analysis.